controller/snapshot_concurrent_limiter.go

```diff
@@ -0,0 +1,178 @@
+package controller
+
+import (
+	"github.com/cockroachdb/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/engineapi"
+	"github.com/longhorn/longhorn-manager/types"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+type snapshotHeavyTask string
+
+var (
+	snapshotPurgeTaskType snapshotHeavyTask = "purge"
+	snapshotCloneTaskType snapshotHeavyTask = "clone"
+)
+
+type SnapshotConcurrentLimiter struct {
+	logger logrus.FieldLogger
+}
+
+// NewSnapshotConcurrentLimiter creates a new SnapshotConcurrentLimiter instance
+// to manage concurrent snapshot heavy I/O tasks like purge and clone.
+func NewSnapshotConcurrentLimiter() *SnapshotConcurrentLimiter {
+	return &SnapshotConcurrentLimiter{
+		logger: logrus.WithField("component", "snapshot-concurrent-limiter"),
+	}
+}
+
+// CanStartSnapshotPurge ensures purge operations do not exceed the concurrent I/O limit.
+func (sc *SnapshotConcurrentLimiter) CanStartSnapshotPurge(engineClientProxy engineapi.EngineClientProxy, engine *longhorn.Engine, ds *datastore.DataStore) (bool, error) {
+	if types.IsDataEngineV2(engine.Spec.DataEngine) {
+		return true, nil
+	}
+
+	return sc.canStartHeavyIOTask(engineClientProxy, engine, ds, snapshotPurgeTaskType)
+}
+
+// CanStartSnapshotClone ensures clone operations do not exceed the concurrent I/O limit.
+func (sc *SnapshotConcurrentLimiter) CanStartSnapshotClone(engineClientProxy engineapi.EngineClientProxy, engine *longhorn.Engine, ds *datastore.DataStore) (bool, error) {
+	return sc.canStartHeavyIOTask(engineClientProxy, engine, ds, snapshotCloneTaskType)
+}
+
+func (sc *SnapshotConcurrentLimiter) canStartHeavyIOTask(engineClientProxy engineapi.EngineClientProxy, engine *longhorn.Engine, ds *datastore.DataStore, taskType snapshotHeavyTask) (bool, error) {
+	switch taskType {
+	case snapshotPurgeTaskType, snapshotCloneTaskType:
+	default:
+		return false, errors.Errorf("unknown snapshot heavy I/O task type: %v", taskType)
+	}
+
+	concurrentLimit, err := ds.GetSettingAsInt(types.SettingNameSnapshotHeavyTaskConcurrentLimit)
+	if err != nil {
+		sc.logger.WithError(err).Debug("Failed to get snapshot purge concurrent limit setting, allowing by default")
+		return true, nil
+	}
+
+	// Unlimited
+	if concurrentLimit < 1 {
+		return true, nil
+	}
+
+	engines, err := ds.ListEngines()
+	if err != nil {
+		return false, err
+	}
+
+	replicas, err := ds.ListReplicas()
+	if err != nil {
+		return false, err
+	}
+
+	// Compute how many heavy tasks are running on each node
+	nodeLoad := make(map[string]int)
+	for _, e := range engines {
+		if !sc.checkEngineBusy(e, engineClientProxy, ds) {
+			continue
+		}
+		sc.iterateReplicasForEngine(e, replicas, nodeLoad)
+	}
+
+	// Determine the nodes that *would be involved* if we start this engine’s task
```

Comment:
It's better to have a comment that describes the race condition. This is a best-effort limiter.